File: cs25backend/src/main/java/edu/dongguk/cs25backend/security/CustomUserDetailService.java
package edu.dongguk.cs25backend.security;

import edu.dongguk.cs25backend.repository.CustomerRepository;
import lombok.RequiredArgsConstructor;
import lombok.extern.slf4j.Slf4j;
import org.springframework.security.core.authority.SimpleGrantedAuthority;
import org.springframework.security.core.userdetails.UserDetails;
import org.springframework.security.core.userdetails.UsernameNotFoundException;
import org.springframework.security.oauth2.client.userinfo.DefaultOAuth2UserService;
import org.springframework.stereotype.Service;

import java.util.ArrayList;
import java.util.Collection;

@Service
@RequiredArgsConstructor
@Slf4j
public class CustomUserDetailService extends DefaultOAuth2UserService {

    private final CustomerRepository customerRepository;

//    @Override
    public UserDetails loadUserByUsername(String username) throws UsernameNotFoundException {

        Collection<SimpleGrantedAuthority> roles = new ArrayList<SimpleGrantedAuthority>();
        roles.add(new SimpleGrantedAuthority("ROLE_USER"));

        UserLoginForm user = customerRepository.findByIdAndRefreshToken(Long.valueOf(username))
                .orElseThrow(() -> new UsernameNotFoundException("ACCESS_DENIED_ERROR"));

        return CustomUserDetail.create(user);
    }

    public UserDetails loadUserByUsernameAndUserRole(String username, String userRole) throws UsernameNotFoundException {
        Collection<SimpleGrantedAuthority> roles = new ArrayList<SimpleGrantedAuthority>();
        roles.add(new SimpleGrantedAuthority(userRole));

        switch (userRole) {
            case "ROLE_CUSTOMER" -> { log.info("CUSTOMER"); }
            case "ROLE_MANAGER" -> { log.info("MANAGER"); }
            case "ROLE_HQ" -> { log.info("HQ"); }

        }

        UserLoginForm user = customerRepository.findByIdAndRefreshToken(Long.valueOf(username))
                .orElseThrow(() -> new UsernameNotFoundException("ACCESS_DENIED_ERROR"));

        return CustomUserDetail.create(user);
    }
}

File: cs25server/src/main/kotlin/edu/dongguk/cs25server/security/CustomUserDetailService.kt
package edu.dongguk.cs25server.security

import edu.dongguk.cs25server.domain.type.UserRole
import edu.dongguk.cs25server.exception.ErrorCode
import edu.dongguk.cs25server.exception.GlobalException
import edu.dongguk.cs25server.repository.CustomerRepository
import edu.dongguk.cs25server.repository.HeadquartersRepository
import edu.dongguk.cs25server.repository.ManagerRepository
import edu.dongguk.cs25server.util.Log.Companion.log
import org.springframework.security.core.authority.SimpleGrantedAuthority
import org.springframework.security.core.userdetails.UserDetails
import org.springframework.security.core.userdetails.UserDetailsService
import org.springframework.stereotype.Service
import java.util.*
import kotlin.collections.ArrayList

@Service
class CustomUserDetailService (
    private val customerRepository: CustomerRepository,
    private val managerRepository: ManagerRepository,
    private val headquartersRepository: HeadquartersRepository
) : UserDetailsService {

    override fun loadUserByUsername(username: String): UserDetails {
        val roles: MutableCollection<SimpleGrantedAuthority> = ArrayList()
        roles.add(SimpleGrantedAuthority("ROLE_USER"))

        val user: UserLoginForm = customerRepository.findByIdAndRefreshToken(username.toLong())
            ?: throw GlobalException(ErrorCode.ACCESS_DENIED_ERROR)

        return CustomUserDetail.create(user)
    }

    fun loadUserByUsernameAndUserRole(username: String, role: String): CustomUserDetail {
        val roles: MutableCollection<SimpleGrantedAuthority> = ArrayList()
        roles.add(SimpleGrantedAuthority("ROLE_$role"))

        val user: UserLoginForm = when (role) {
            "CUSTOMER" -> {
                customerRepository.findByIdAndRefreshToken(username.toLong())
                    ?: throw GlobalException(ErrorCode.ACCESS_DENIED_ERROR)
            }
            "MANAGER" -> {
                managerRepository.findByIdAndRefreshToken(username.toLong())
                    ?: throw GlobalException(ErrorCode.ACCESS_DENIED_ERROR)
            }
            "HQ" -> {
                headquartersRepository.findByIdAndRefreshToken(username.toLong())
                    ?: throw GlobalException(ErrorCode.ACCESS_DENIED_ERROR)
            }
            else -> { throw GlobalException(ErrorCode.ACCESS_DENIED_ERROR)}
        }

        return CustomUserDetail.create(user)
    }
}

